<a href="https://colab.research.google.com/github/RuthNduta/Data-Visualization-with-Dash-Plotly/blob/main/template_notebook_data_analysis_and_visualization_101_bootcamp_final_assignment_f54c706.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Overview
Throughout this assignment, you will be performing certain well-defined tasks that’ll not only strengthen your concepts of Plotly and Dash, but will also help you learn a number of new concepts that are useful in analyzing, summarizing and visualizing data in the real world. 

Here is a template notebook with all the tasks mentioned in detail. **Please complete the tasks within the designated section only.**


## Task 1: Data Loading and Data Aggregation
* Load the 3 data files into the variables data_18, data_19, data_20. 

* Data aggregation is the process of gathering data and presenting it in a summarized format. The data may be gathered from multiple data sources with the intent of combining these data sources into a summary for data analysis.         
Since this dataset involves 3 data files, you’ll often be working on combining information from 2 or more files and analysing it. More often than not, GroupBy is a very useful tool for this purpose. 

Go through this article to learn more some helpful aggregation tools in Python: https://www.bmc.com/blogs/pandas-group-merge-concatenate-join/ 

In [1]:
#Importing libraries to be used

import pandas as pd
import numpy as np
import matplotlib
from matplotlib import pyplot as plt

In [2]:
data18 = pd.read_csv("https://raw.githubusercontent.com/dphi-official/Datasets/master/IT_Salary_Survey_EU_18-20/Survey_2018.csv")
data19 = pd.read_csv("https://raw.githubusercontent.com/dphi-official/Datasets/master/IT_Salary_Survey_EU_18-20/Survey_2019.csv")
data20 = pd.read_csv("https://raw.githubusercontent.com/dphi-official/Datasets/master/IT_Salary_Survey_EU_18-20/Survey_2020.csv")

In [3]:
data18.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 765 entries, 0 to 764
Data columns (total 14 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Timestamp                           765 non-null    object 
 1   Age                                 672 non-null    float64
 2   Gender                              751 non-null    object 
 3   City                                736 non-null    object 
 4   Position                            737 non-null    object 
 5   Years of experience                 732 non-null    float64
 6   Your level                          743 non-null    object 
 7   Current Salary                      750 non-null    float64
 8   Salary one year ago                 596 non-null    float64
 9   Salary two years ago                463 non-null    float64
 10  Are you getting any Stock Options?  742 non-null    object 
 11  Main language at work               750 non-n

In [4]:
data19.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 991 entries, 0 to 990
Data columns (total 23 columns):
 #   Column                                                                                                Non-Null Count  Dtype  
---  ------                                                                                                --------------  -----  
 0   Zeitstempel                                                                                           991 non-null    object 
 1   Age                                                                                                   882 non-null    float64
 2   Gender                                                                                                991 non-null    object 
 3   City                                                                                                  991 non-null    object 
 4   Seniority level                                                                                       97

## Task 2: Data Analysis
* Display the first 5 rows of the 2018 survey data
* Display a concise summary of the 2020 data and list out 3 observations/inferences that you observe from the result. For this you will need to use the info() method.
* Display the descriptive statistics of the 2018 survey data
* Display the number of missing values in each column of the 2018 survey data
How many people responded to the survey in each of the 3 years? Has the number increased or decreased over the years?
* Display all the unique values and their frequency in the column - “Number of vacation days” of 2020 data. Write down your observations (at least one) for this result. 


In [5]:
#Display the first 5 rows of the 2018 survey data
data18.head()

Timestamp   Age  ... Company size   Company type
0  14/12/2018 12:41:33  43.0  ...     100-1000        Product
1  14/12/2018 12:42:09  33.0  ...       50-100        Product
2  14/12/2018 12:47:36  32.0  ...        1000+        Product
3  14/12/2018 12:50:15  25.0  ...        1000+        Product
4  14/12/2018 12:50:31  39.0  ...     100-1000  Ecom retailer

[5 rows x 14 columns]

In [6]:
#Display a concise summary of the 2020 data and list out 3 observations/inferences that you observe from the result. For this you will need to use the info() method.
data20.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1253 entries, 0 to 1252
Data columns (total 23 columns):
 #   Column                                                                                                                   Non-Null Count  Dtype  
---  ------                                                                                                                   --------------  -----  
 0   Timestamp                                                                                                                1253 non-null   object 
 1   Age                                                                                                                      1226 non-null   float64
 2   Gender                                                                                                                   1243 non-null   object 
 3   City                                                                                                                     1253 non-null   o

From the 2020 data we can observe that: 
    1. It has 23 columns and 1253 rows
    2. The columns have many anomalies
    3. Most of the data types are objects

In [7]:
#Display the descriptive statistics of the 2018 survey data
data18.describe()

Age  ...  Salary two years ago
count  672.000000  ...            463.000000
mean    32.183036  ...          58013.475162
std      5.107268  ...          20413.048908
min     21.000000  ...          10001.000000
25%     29.000000  ...          48000.000000
50%     32.000000  ...          56000.000000
75%     35.000000  ...          67000.000000
max     60.000000  ...         150000.000000

[8 rows x 5 columns]

In [8]:
#Display the number of missing values in each column of the 2018 survey data 
data18.isnull().sum()

Timestamp                               0
Age                                    93
Gender                                 14
City                                   29
Position                               28
Years of experience                    33
Your level                             22
Current Salary                         15
Salary one year ago                   169
Salary two years ago                  302
Are you getting any Stock Options?     23
Main language at work                  15
Company size                           15
Company type                           35
dtype: int64

In [9]:
#How many people responded to the survey in each of the 3 years? Has the number increased or decreased over the years?
print("2018:", len(data18))
print("2019:", len(data19))
print("2020:", len(data20))



2018: 765
2019: 991
2020: 1253


In [10]:
#Display all the unique values and their frequency in the column - “Number of vacation days” of 2020 data. Write down your observations (at least one) for this result.
print("unique values and their frequency in the column:","\n", data20['Number of vacation days'].value_counts())
print("*"*10)
print("Total number of unique values is:", data20['Number of vacation days'].nunique())

unique values and their frequency in the column: 
 30                                              488
28                                              233
27                                              102
25                                               91
26                                               71
24                                               67
29                                               24
20                                               13
21                                               10
32                                                8
22                                                8
31                                                8
35                                                5
36                                                5
0                                                 4
40                                                4
23                                                4
unlimited                                         3
45           

## Task 3: Data Cleaning
* Rename the column ‘Position ‘ in the 2020 data to ‘Position’. (without the blank space)
* Check for missing values in 2020 data for all the columns. If there are no missing values, proceed to the next step. If there are missing values in the dataset,
  * For categorical variables, fill the missing values with the mode of the data. Remember if the data type of any variable is ‘object’, it is categorical variable. 
  * For numerical variables, fill the missing values with the mean of the data.

Here's a good blog that displays multiple methods of filling (imputing) missing values: https://jamesrledoux.com/code/imputation 
* Drop the timestamp column for all the three years data since the date and time at which a person filled the survey is irrelevant to us. The year matters and we already know that from the dataset’s name.
* Perform any other data cleaning steps you believe are necessary. (removing outliers, handling missing values in a way to beautify visualizations, making the categories uniform i.e python and Python should mean the same thing etc.) Note that the same steps will have to be performed for all 3 data files.

In [11]:
#Rename the column ‘Position ‘ in the 2020 data to ‘Position’. (without the blank space)
data20['Position '] = data20['Position '].str.strip()

In [12]:
#Check for missing values in 2020 data for all the columns.

data20.isnull().sum().sort_values(ascending = False)

Have you been forced to have a shorter working week (Kurzarbeit)? If yes, how many hours per week                          880
Have you received additional monetary support from your employer due to Work From Home? If yes, how much in 2020 in EUR    791
Annual bonus+stocks one year ago. Only answer if staying in same country                                                   639
Yearly bonus + stocks in EUR                                                                                               424
Annual brutto salary (without bonus and stocks) one year ago. Only answer if staying in the same country                   368
Other technologies/programming languages you use often                                                                     157
Your main technology / programming language                                                                                127
Number of vacation days                                                                                        

In [13]:
survey = pd.concat([data18, data19, data20], axis=0, join='inner') 
survey.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3009 entries, 0 to 1252
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Age                    2780 non-null   float64
 1   Gender                 2985 non-null   object 
 2   City                   2980 non-null   object 
 3   Main language at work  2973 non-null   object 
 4   Company size           2962 non-null   object 
 5   Company type           2918 non-null   object 
dtypes: float64(1), object(5)
memory usage: 164.6+ KB


In [14]:
#Imputing missing values
survey['Age'].fillna(survey['Age'].mean(), inplace=True)
survey['Gender'].fillna(survey['Gender'].mode()[0], inplace=True)
survey['City'].fillna(survey['City'].mode()[0], inplace=True)
survey['Main language at work'].fillna(survey['Main language at work'].mode()[0], inplace=True)
survey['Company size'].fillna(survey['Company size'].mode()[0], inplace=True)
survey['Company type'].fillna(survey['Company type'].mode()[0], inplace=True)

In [15]:
survey.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3009 entries, 0 to 1252
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Age                    3009 non-null   float64
 1   Gender                 3009 non-null   object 
 2   City                   3009 non-null   object 
 3   Main language at work  3009 non-null   object 
 4   Company size           3009 non-null   object 
 5   Company type           3009 non-null   object 
dtypes: float64(1), object(5)
memory usage: 164.6+ KB


## Task 4: Data Visualization using Plotly
**Note:** All the tasks below need to be completed using only Plotly and no other Data Visualization library.

* Create a pie chart to analyze the Company types in the year 2019. Are Consulting / Agency companies more popular than Startups? 
* Create a line plot of the Total years of experience vs the current salary(taking the median salary for each of the different experience years) of the year 2018.
* Now, create the above plot again and add 2 more line plots to the same graph, that display the Total years of experience vs the median Yearly brutto salary (without bonus and stocks) of the year 2019 and 2020.
* Create a bar chart to analyse the popularity of the main technology/ programming languages amongst the respondents in the year 2020. Which technology is the most popular? Which technology is the least popular (with less than 4 responses)?
* Create a pie plot indicating the gender ratio of the respondents in the year 2020.


In [16]:
#Create a pie chart to analyze the Company types in the year 2019. 
#Are Consulting / Agency companies more popular than Startups?
data19.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 991 entries, 0 to 990
Data columns (total 23 columns):
 #   Column                                                                                                Non-Null Count  Dtype  
---  ------                                                                                                --------------  -----  
 0   Zeitstempel                                                                                           991 non-null    object 
 1   Age                                                                                                   882 non-null    float64
 2   Gender                                                                                                991 non-null    object 
 3   City                                                                                                  991 non-null    object 
 4   Seniority level                                                                                       97

In [17]:
data19['Company type'].astype('category')
data19['Company type'].fillna(data19['Company type'].mode()[0], inplace=True)


In [18]:
#Importing relevant libraries
import plotly.express as px
# import graph objects as "go"
import plotly.graph_objs as go
#import plotly.plotly
#!pip install chart-studio
import chart_studio.plotly as plotly
from chart_studio.plotly import plot, iplot
from plotly.offline import iplot

In [42]:
#Creating  a pie chart

fig = px.pie(data19,names='Company type', hole=0.3) #customize it to a donut shape instead of a pie using the hole parameter
fig.update_traces(textinfo="label+percent", insidetextfont= dict(color="white")) #Changing the label & colour of the labels
#fig.update_layout(legend = {"itemclick": False}) #Changing what happens on clicking on a legend
fig.show()

Create a line plot of the Total years of experience vs the current salary(taking the median salary for each of the different experience years) of the year 2018.

In [21]:
#Preparing data from 2018:
data18['Years of experience'].fillna(data18['Years of experience'].mean(), inplace= True)
data18['Current Salary'].fillna(data18['Current Salary'].mean(), inplace= True)

In [32]:
#Grouping the 2018 data

df = data18[['Years of experience', 'Current Salary']]
df2 = df.groupby(['Years of experience'],as_index=False)['Current Salary'].median()
df2.head()

Years of experience  Current Salary
0                  0.0         50000.0
1                  0.5         46000.0
2                  1.0         46000.0
3                  1.5         58000.0
4                  2.0         50000.0

In [107]:
#Line Plot
# Creating trace1
trace1 = go.Scatter(
                    x = df2['Years of experience'],
                    y = df2['Current Salary'],
                    mode = "lines",
                    name = "2018",
                    marker = dict(color = 'rgba(16, 112, 2, 0.8)'),
                    text= data18['Company type'])

In [108]:
data = [trace1]
layout = dict(title = 'Total years of experience vs the current salary',
              xaxis= dict(title= 'years of experience',ticklen= 5,zeroline= False)
             )
fig = dict(data = data, layout = layout)
iplot(fig)

In [109]:
px.line(df2, x = df2['Years of experience'],y = df2['Current Salary'])

Now, create the above plot again and add 2 more line plots to the same graph, that display the Total years of experience vs the median Yearly brutto salary (without bonus and stocks) of the year 2019 and 2020.


In [ ]:
data19.info()

In [125]:
df = data19[['Years of experience', 'Yearly brutto salary (without bonus and stocks)']]
df['Yearly brutto salary (without bonus and stocks)'].fillna(df['Yearly brutto salary (without bonus and stocks)'].mean(), inplace= True)
df19 = df.groupby(['Years of experience'],as_index=False)['Yearly brutto salary (without bonus and stocks)'].median()
df19.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26 entries, 0 to 25
Data columns (total 2 columns):
 #   Column                                           Non-Null Count  Dtype  
---  ------                                           --------------  -----  
 0   Years of experience                              26 non-null     int64  
 1   Yearly brutto salary (without bonus and stocks)  26 non-null     float64
dtypes: float64(1), int64(1)
memory usage: 624.0 bytes


/usr/local/lib/python3.7/dist-packages/pandas/core/series.py:4536: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
data20

In [149]:
#Preparing 2020 data

df20 = data20[['Total years of experience', 'Yearly brutto salary (without bonus and stocks) in EUR']]
df20 = df20.groupby(['Total years of experience'],as_index=False)['Yearly brutto salary (without bonus and stocks) in EUR'].median()
df20.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48 entries, 0 to 47
Data columns (total 2 columns):
 #   Column                                                  Non-Null Count  Dtype  
---  ------                                                  --------------  -----  
 0   Total years of experience                               48 non-null     object 
 1   Yearly brutto salary (without bonus and stocks) in EUR  48 non-null     float64
dtypes: float64(1), object(1)
memory usage: 1.1+ KB


In [84]:
#Removing missing values

df['Total years of experience'].fillna(df['Total years of experience'].mode(), inplace= True)
df['Yearly brutto salary (without bonus and stocks) in EUR'].fillna(df['Yearly brutto salary (without bonus and stocks) in EUR'].mean(), inplace= True)


/usr/local/lib/python3.7/dist-packages/pandas/core/series.py:4536: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [103]:
# Creating trace2
trace2 = go.Scatter(
                    x = df19['Years of experience'],
                    y = df19['Yearly brutto salary (without bonus and stocks)'],
                    mode = "lines+markers",
                    name = "2019",
                    marker = dict(color = 'rgba(80, 26, 80, 0.8)'),
                    text= data19['Company type'])
data = [trace1, trace2]
layout = dict(title = 'Years of experience vs Yearly brutto salary (without bonus and stocks) in 2019 & Current salary in 2020',
              xaxis= dict(title= 'Years of experience',ticklen= 5,zeroline= False)
             )
fig = dict(data = data, layout = layout)
iplot(fig)

In [151]:
data20['Total years of experience'].astype('category')
df20['Total years of experience']


0                                                     0
1                                                   0.8
2                                                     1
3                      1 (as QA Engineer) / 11 in total
4                                                   1,5
5                                                   1.5
6                                                    10
7                                                    11
8                                                    12
9                                                    13
10                                                   14
11                                                   15
12                                 15, thereof 8 as CTO
13                                                   16
14                                                   17
15                                                   18
16                                                   19
17                                              

In [ ]:
df20['Total years of experience']


In [110]:
# Creating trace3
trace2 = go.Scatter(
                    x = df20['Total years of experience'],
                    y = df20['Yearly brutto salary (without bonus and stocks) in EUR'],
                    mode = "lines+markers",
                    name = "2019",
                    marker = dict(color = 'rgba(80, 26, 80, 0.8)'),
                     text= data20['Company type'])
                
data = [trace1, trace2, trace3]
layout = dict(title = 'Years of experience vs Yearly brutto salary (without bonus and stocks) in 2019, 2020 & Current salary in 2020',
              xaxis= dict(title= 'Salary Trends',ticklen= 5,zeroline= False),
              yaxis = dict(title='Salary', ticklen=5, zeroline= False)
             )
fig = dict(data = data, layout = layout)
iplot(fig)

Create a bar chart to analyse the popularity of the main technology/ programming languages amongst the respondents in the year 2020. Which technology is the most popular? 


In [168]:
df20b = data20['Your main technology / programming language'].value_counts().head()
px.bar(df20b)


Which technology is the least popular (with less than 4 responses)?

In [169]:
df20c = data20['Your main technology / programming language'].value_counts().tail()
px.bar(df20c)

Create a pie plot indicating the gender ratio of the respondents in the year 2020.

In [185]:
data20['Gender'].fillna(data20['Gender'].mode()[0], inplace=True)
df20d = data20['Gender'].value_counts()

fig = px.pie(df20d,names=data20['Gender']) #customize it to a donut shape instead of a pie using the hole parameter
fig.update_traces(textinfo="label+percent", insidetextfont= dict(color="white")) #Changing the label & colour of the labels
#fig.update_layout(legend = {"itemclick": False}) #Changing what happens on clicking on a legend
fig.show()

## Bonus Section [Optional but carries bonus marks]
This dataset is as raw and real as it can get while conducting yearly surveys. You might have observed that the data is not clean and structured and requires some thorough cleaning before deriving meaningful plots. When combined with the power of Plotly and Dash, there are endless possibilities for the insightful visualizations you can create. 

This section is to let you experiment, explore and create as many visualizations as you’d like. You never know, if we like the creativity and the extra work, you might receive some bonus marks!


In [26]:
#!pip install dash==0.21.1  
#!pip install dash-renderer==0.13.0  
#!pip install dash-html-components==0.11.0
#!pip install dash-core-components==0.23.0  
#!pip install plotly --upgrade

In [27]:
import dash
import dash_core_components as dcc
import dash_html_components as html

In [28]:
app = dash.Dash()
colors = {
    'background': '#111111',
    'text': '#7FDBFF'
}
app.layout = html.Div(style={'backgroundColor': colors['background']}, children=[
    html.H1(
        children='Hello Dash',
        style={
            'textAlign': 'center',
            'color': colors['text']
        }
    ),
    html.Div(children='Dash: A web application framework for Python.', style={
        'textAlign': 'center',
        'color': colors['text']
    }),
    dcc.Graph(
        id='Graph1',
        figure={
            'data' = [{trace1}],
            'layout': {
                'plot_bgcolor': colors['background'],
                'paper_bgcolor': colors['background'],
                'font': {
                    'color': colors['text']
                    'title' :{} 'Total years of experience vs the current salary',
                    
layout = dict(
              xaxis= dict(title= 'years of experience',ticklen= 5,zeroline= False)
             )
fig = dict(data = data, layout = layout)
iplot(fig)
                }
            }
        }
    )
])

In [29]:
if __name__ == '__main__':
    app.run_server(debug=True)

 * Serving Flask app "__main__" (lazy loading)
 * Environment: production
   Use a production WSGI server instead.
 * Debug mode: on


 * Running on http://127.0.0.1:8050/ (Press CTRL+C to quit)
 * Restarting with stat


SystemExit: ignored

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2890: UserWarning:

To exit: use 'exit', 'quit', or Ctrl-D.



# Conclusion
This brings us to the end of the assignment and to the bootcamp. We hope you had a great learning time. :)

Now, you can submit your notebook for assessment. 